design/permissions.md

```diff
@@ -0,0 +1,219 @@
+7. [Runtime] Determine if the MRE has sufficient permission to run, i.e. `execution` and all the required permissions
+	are granted. If not, abort connecting to the MRE.
+8. [Runtime] Startup proceeds like normal, but if a `user-joined` message is sent, the user payload must include
+	the IDs of any permissions that were granted to this MRE by this user (other than `execution`, which is implied).
+
+
+Permissions Manager
+---------------------
+
+Host apps must provide hooks for the MRE subsystem to obtain permission from users. This is provided to
+`MREAPI.InitializeAPI` as an implementation of `IPermissionManager`, which has the following methods:
+
+* `Task<IEnumerable<Permissions>> PromptForPermission(...)` - Request permissions from the user, and return a Task that
+	resolves with those permissions the user has granted. Takes the following arguments:
+	* `string appOrigin` - The origin of the MRE requesting permission, as described above.
+	* `string appDisplayName` - A human-readable identifier for the MRE server provided from the manifest.
+	* `IEnumerable<Permissions> permissionsNeeded` - A list of the permissions required to run the app.
+	* `IEnumerable<Permissions> permissionsWanted` - A list of permissions that the app can use, but are not
+		needed to run.
+* `IEnumerable<Permissions> CurrentPermissions(...)` - Returns a list of the currently granted permissions. Takes
+	the following arguments:
+	* `string appOrigin` - The origin of the MRE for which we want the permission set.
+
+The `Permissions` enum has the `Flags` attribute and is on powers of 2 in case testing combinations is desired.
+
+```cs
+[Flags]
+enum Permissions {
+	None = 0,
+	Execution = 1,
+	UserTracking = 2,
+	UserInteraction = 4,
+	...
+}
+```
+
+The default implementation of the permissions manager will also include the following virtual methods:
+
+* `void ReadFromStorage()` - Populate the in-memory permission database from offline storage.
+* `void WriteToStorage()` - Write the "remembered" portions of the in-memory permission database to offline storage.
+
+
+UI Mockups
+============
+
+This is what an MRE permissions management interface could look like. When an MRE is available to run in the current
+host app context, a button to bring up the management UI should be readily available (within two clicks of steady-state
+usage). Attention should be drawn to this button when a new MRE permission request comes in.
+
+Clicking the prompt or the button brings up your MRE permission requests:
+
+![Requests](images/permission.requests.png)
+
+This window is paginated, one page per MRE that requires permissions. The name of the app requesting permission, the
+server name it's hosted on, and a secure protocol indicator are all featured prominently. Each permission has a
+checkbox next to it: required permissions are checked and disabled, and optional permissions are checked or not based
+on app origin decision history and settings.
```

Comment:
Suggestion: indicate the number of pending requests in the tab label, like the friends requests UI does